slider/curve.py

```diff
@@ -128,13 +123,42 @@ def __call__(self, t):
         return self._curves[bi]((t - pre_t) / (post_t - pre_t))
 
 
-class Passthrough(Curve):
-    kinds = 'P'
+class LinearMetaCurve(MetaCurve):
+    kinds = 'L'
 
     def __init__(self, points, req_length):
         self.points = points
         self.req_length = req_length
-        self._center = center = get_center(*self.points)
+        self._curves = [
+            Bezier(subpoints, None) for subpoints in sliding_window(2, points)
+        ]
+
+
+class Perfect(Curve):
+    kinds = 'P'
+
+    def __new__(cls, points, req_length):
+        if len(points) != 3:
+            # it seems osu! uses the bezier curve if there are more than 3
+            # points
+            # https://github.com/ppy/osu/blob/7fbbe74b65e7e399072c198604e9db09fb729626/osu.Game/Rulesets/Objects/SliderCurve.cs#L32  # noqa
+            return Bezier(points, req_length)
+
+        try:
+            center = get_center(*points)
+        except ValueError:
+            # we cannot use a perfect curve function for collinear points;
+            # osu! also falls back to a bezier here
+            return Bezier(points, req_length)
+
+        self = super().__new__(cls)
+        self._init(points, req_length, center)
+        return self
+
+    def _init(self, points, req_length, center):
+        self.points = points
+        self.req_length = req_length
+        self._center = center
```

Comment:
I just realized that this isn't being flipped in HR. I think we should just add a `hard_rock` attribute here to compute this so the code in `Slider` can just call that.